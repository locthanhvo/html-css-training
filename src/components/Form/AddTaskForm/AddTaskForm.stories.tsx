import { StoryObj, Meta } from '@storybook/react';
import { BrowserRouter } from 'react-router-dom';

import AddTaskForm from '.';

const meta: Meta<typeof AddTaskForm> = {
  title: 'Components/AddTaskForm',
  tags: ['autodocs'],
  component: AddTaskForm,
  decorators: [
    (Story) => (
      <BrowserRouter>
        <Story />
      </BrowserRouter>
    ),
  ],
  parameters: {
    layout: 'centered',
  },
};

export default meta;
type Story = StoryObj<typeof AddTaskForm>;

export const Default: Story = {
  args: {},
};
