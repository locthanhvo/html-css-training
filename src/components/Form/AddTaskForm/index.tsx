import { memo, useCallback } from 'react';
import { Button, Card, Flex, FormControl } from '@chakra-ui/react';
import { Controller, useForm } from 'react-hook-form';

// Types
import { IInputField, TAddTaskForm } from '@/types';

// Constants
import { TASK_SCHEMA } from '@/constants';

// Components
import InputField from '@/components/common/InputField';
import DatePicker from '@/components/common/DatePicker';
import UploadFile from '@/components/UploadFile';

// Icons
import { CloseIcon, FlagIcon } from '@/components/common/Icons';

// Hooks
import { useUploader } from '@/hooks';

interface AddTaskFormProps {
  isLoading?: boolean;
  onSubmit?: (data: TAddTaskForm, imageFiles?: File[]) => void;
  onClose?: () => void;
}

const AddTaskForm = ({
  isLoading = false,
  onSubmit,
  onClose,
}: AddTaskFormProps) => {
  const { previewURLs, imageFiles, getRootProps, getInputProps } =
    useUploader();

  const {
    control,
    formState: { isSubmitting, isValid },
    handleSubmit,
    clearErrors,
  } = useForm<TAddTaskForm>({
    mode: 'onBlur',
    reValidateMode: 'onChange',
    defaultValues: {
      title: '',
      startDate: '',
      endDate: '',
    },
  });

  const handleInputChange = useCallback(
    ({ field, data, isError, onChange }: IInputField<TAddTaskForm>) => {
      isError && clearErrors(field);
      onChange(data);
    },
    [clearErrors],
  );

  const handleSubmitForm = useCallback(
    (data: TAddTaskForm) => {
      onSubmit?.(data, imageFiles);
    },
    [onSubmit, imageFiles],
  );

  return (
    <Card
      w={230}
      as="form"
      bg="white"
      p={4}
      gap={5}
      direction="column"
      borderRadius="md"
      onSubmit={handleSubmit(handleSubmitForm)}
    >
      <Flex gap={2} alignItems="center">
        <Button
          variant="outline"
          rightIcon={<CloseIcon />}
          onClick={onClose}
          isDisabled={isLoading || isSubmitting}
        />

        <Controller
          name="title"
          rules={TASK_SCHEMA.TITLE}
          control={control}
          render={({ field, fieldState: { error } }) => (
            <InputField
              fontSize="base"
              variant="outline"
              {...field}
              isDisabled={isLoading || isSubmitting}
              onChange={(data) =>
                handleInputChange({
                  field: 'title',
                  data,
                  isError: !!error,
                  onChange: field.onChange,
                })
              }
              type="text"
              placeholder="Task name or type "
            />
          )}
        />
      </Flex>

      <Flex justifyContent="space-between" alignItems="center">
        <Flex gap={1}>
          <Button variant="outline" rightIcon={<FlagIcon />} />

          <Controller
            name="images"
            control={control}
            render={() => (
              <FormControl>
                <UploadFile
                  isDisabled={isLoading || isSubmitting}
                  previewURLs={previewURLs}
                  getInputProps={getInputProps}
                  getRootProps={getRootProps}
                />
              </FormControl>
            )}
          />

          <FormControl>
            <DatePicker
              control={control}
              isDisabled={isLoading || isSubmitting}
              onChange={handleInputChange}
            />
          </FormControl>
        </Flex>

        <Button
          type="submit"
          variant="primary"
          isDisabled={!isValid && !isSubmitting}
          isLoading={isLoading || isSubmitting}
        >
          Save
        </Button>
      </Flex>
    </Card>
  );
};

export default memo(AddTaskForm);
