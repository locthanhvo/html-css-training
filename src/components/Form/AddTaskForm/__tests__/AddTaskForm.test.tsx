import { fireEvent, render, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import AddTaskForm from '..';

const mockOnSubmit = jest.fn();
const mockOnClose = jest.fn();

describe('AddTaskForm component', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should match snapshot', () => {
    const { container } = render(
      <AddTaskForm onSubmit={mockOnSubmit} onClose={mockOnClose} />,
    );
    expect(container).toMatchSnapshot();
  });

  it('should call onSubmit', async () => {
    const { getByRole, getByPlaceholderText } = render(
      <AddTaskForm
        onSubmit={mockOnSubmit}
        onClose={mockOnClose}
        isLoading={false}
      />,
    );

    const taskInput = getByPlaceholderText('Task name or type');
    fireEvent.change(taskInput, { target: { value: 'Company ABC' } });

    await userEvent.click(getByRole('button', { name: 'Save' }));

    await waitFor(() => expect(mockOnSubmit).toHaveBeenCalled());
  });
});
