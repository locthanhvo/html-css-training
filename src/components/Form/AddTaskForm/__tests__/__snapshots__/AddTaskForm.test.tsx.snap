// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddTaskForm component should match snapshot 1`] = `
<div>
  <form
    class="chakra-card css-13rbcq2"
  >
    <div
      class="css-r2hvuq"
    >
      <button
        class="chakra-button css-4xx2wk"
        type="button"
      >
        <span
          class="chakra-button__icon css-1hzyiq5"
        >
          <svg
            fill="none"
            height="31"
            viewBox="0 0 30 31"
            width="30"
          >
            <path
              clip-rule="evenodd"
              d="M20.8925 8.80249C20.7057 8.61523 20.452 8.51 20.1875 8.51C19.923 8.51 19.6693 8.61523 19.4825 8.80249L14.5925 13.6825L9.7025 8.79249C9.51567 8.60523 9.26202 8.5 8.9975 8.5C8.73298 8.5 8.47933 8.60523 8.2925 8.79249C7.9025 9.18249 7.9025 9.81249 8.2925 10.2025L13.1825 15.0925L8.2925 19.9825C7.9025 20.3725 7.9025 21.0025 8.2925 21.3925C8.6825 21.7825 9.3125 21.7825 9.7025 21.3925L14.5925 16.5025L19.4825 21.3925C19.8725 21.7825 20.5025 21.7825 20.8925 21.3925C21.2825 21.0025 21.2825 20.3725 20.8925 19.9825L16.0025 15.0925L20.8925 10.2025C21.2725 9.82249 21.2725 9.18249 20.8925 8.80249Z"
              fill="#C3CAD9"
              fill-rule="evenodd"
            />
          </svg>
        </span>
      </button>
      <div
        class="chakra-form-control css-17qx7uq"
        role="group"
      >
        <div
          class="chakra-input__group css-bx0blc"
          data-group="true"
        >
          <input
            class="chakra-input css-1g2zu81"
            id="field-:r0:"
            name="title"
            placeholder="Task name or type "
            type="text"
            value=""
          />
        </div>
      </div>
    </div>
    <div
      class="css-69i1ev"
    >
      <div
        class="css-nn803e"
      >
        <button
          class="chakra-button css-4xx2wk"
          type="button"
        >
          <span
            class="chakra-button__icon css-1hzyiq5"
          >
            <svg
              fill="none"
              height="31"
              viewBox="0 0 30 31"
              width="30"
            >
              <path
                clip-rule="evenodd"
                d="M17.4 9.5L17.16 8.3C17.07 7.84 16.66 7.5 16.18 7.5H9C8.45 7.5 8 7.95 8 8.5V23.5C8 24.05 8.45 24.5 9 24.5C9.55 24.5 10 24.05 10 23.5V17.5H15.6L15.84 18.7C15.93 19.17 16.34 19.5 16.82 19.5H22C22.55 19.5 23 19.05 23 18.5V10.5C23 9.95 22.55 9.5 22 9.5H17.4Z"
                fill="#C3CAD9"
                fill-rule="evenodd"
              />
            </svg>
          </span>
        </button>
        <div
          class="chakra-form-control css-0"
          role="group"
        >
          <button
            class="chakra-button css-8y5phl"
            role="presentation"
            tabindex="0"
            type="button"
          >
            <div
              class="css-w1wtgb"
            >
              <svg
                fill="none"
                height="31"
                viewBox="0 0 30 31"
                width="30"
              >
                <path
                  clip-rule="evenodd"
                  d="M20.2539 20H9.67393C7.58393 20 5.72393 18.47 5.52393 16.39C5.29393 14.01 7.16393 12 9.50393 12H21.8639C23.1739 12 24.3639 12.94 24.4939 14.24C24.6439 15.74 23.4739 17 22.0039 17H11.5039C10.9539 17 10.5039 16.55 10.5039 16C10.5039 15.45 10.9539 15 11.5039 15H20.2539C20.6639 15 21.0039 14.66 21.0039 14.25C21.0039 13.84 20.6639 13.5 20.2539 13.5H11.6439C10.3339 13.5 9.14393 14.44 9.01393 15.74C8.86393 17.24 10.0339 18.5 11.5039 18.5H21.8339C23.9239 18.5 25.7839 16.97 25.9839 14.89C26.2139 12.5 24.3439 10.5 22.0039 10.5H9.73393C6.86393 10.5 4.29393 12.6 4.02393 15.46C3.72393 18.75 6.28393 21.5 9.50393 21.5H20.2539C20.6639 21.5 21.0039 21.16 21.0039 20.75C21.0039 20.34 20.6639 20 20.2539 20Z"
                  fill="#C3CAD9"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
            <input
              accept="image/jpeg,image/png,image/webp,image/heic,image/jfif"
              class="chakra-input css-1hyfx7x"
              id="field-:r1:"
              multiple=""
              style="display: none;"
              tabindex="-1"
              type="file"
            />
          </button>
        </div>
        <div
          class="chakra-form-control css-0"
          role="group"
        >
          <button
            aria-controls="popover-content-:r4:"
            aria-expanded="false"
            aria-haspopup="dialog"
            class="chakra-button css-8y5phl"
            id="popover-trigger-:r4:"
            type="button"
          >
            <div
              class="css-w1wtgb"
            >
              <svg
                fill="none"
                height="31"
                viewBox="0 0 30 31"
                width="30"
              >
                <path
                  clip-rule="evenodd"
                  d="M18.4064 18.5314L15.0364 16.5314V12.0014C15.0364 11.6014 14.7164 11.2814 14.3164 11.2814H14.2564C13.8564 11.2814 13.5364 11.6014 13.5364 12.0014V16.7214C13.5364 17.0714 13.7164 17.4014 14.0264 17.5814L17.6764 19.7714C18.0164 19.9714 18.4564 19.8714 18.6564 19.5314C18.8664 19.1814 18.7564 18.7314 18.4064 18.5314ZM23.7164 8.29139L20.6364 5.73139C20.2164 5.38139 19.5864 5.43139 19.2264 5.86139C18.8764 6.28139 18.9364 6.91139 19.3564 7.27139L22.4264 9.83139C22.8464 10.1814 23.4764 10.1314 23.8364 9.70139C24.1964 9.28139 24.1364 8.65139 23.7164 8.29139ZM6.63638 9.83139L9.70638 7.27139C10.1364 6.91139 10.1964 6.28139 9.83638 5.86139C9.48638 5.43139 8.85638 5.38139 8.43638 5.73139L5.35638 8.29139C4.93638 8.65139 4.87638 9.28139 5.23638 9.70139C5.58638 10.1314 6.21638 10.1814 6.63638 9.83139ZM14.5364 7.28139C9.56638 7.28139 5.53638 11.3114 5.53638 16.2814C5.53638 21.2514 9.56638 25.2814 14.5364 25.2814C19.5064 25.2814 23.5364 21.2514 23.5364 16.2814C23.5364 11.3114 19.5064 7.28139 14.5364 7.28139ZM14.5364 23.2814C10.6764 23.2814 7.53638 20.1414 7.53638 16.2814C7.53638 12.4214 10.6764 9.28139 14.5364 9.28139C18.3964 9.28139 21.5364 12.4214 21.5364 16.2814C21.5364 20.1414 18.3964 23.2814 14.5364 23.2814Z"
                  fill="#C3CAD9"
                  fill-rule="evenodd"
                />
              </svg>
            </div>
          </button>
          <div
            class="chakra-popover__popper css-0"
            style="visibility: hidden; position: absolute; min-width: max-content; inset: 0 auto auto 0;"
          >
            <section
              aria-describedby="popover-body-:r4:"
              class="chakra-popover__content css-1wnj9b6"
              id="popover-content-:r4:"
              role="dialog"
              style="transform-origin: var(--popper-transform-origin); opacity: 0; visibility: hidden; transform: scale(0.95) translateZ(0);"
              tabindex="-1"
            >
              <div
                class="chakra-popover__body css-0"
                id="popover-body-:r4:"
              >
                <div
                  class="chakra-stack css-1kkt86i"
                >
                  <div
                    class="chakra-form-control css-17qx7uq"
                    role="group"
                  >
                    <label
                      class="chakra-form__label css-zecyy3"
                      for="field-:r5:"
                      id="field-:r5:-label"
                    >
                      Start Date
                    </label>
                    <div
                      class="chakra-input__group css-bx0blc"
                      data-group="true"
                    >
                      <input
                        class="chakra-input css-1lx593h"
                        id="field-:r5:"
                        name="startDate"
                        type="date"
                        value=""
                      />
                    </div>
                  </div>
                  <div
                    class="chakra-form-control css-17qx7uq"
                    role="group"
                  >
                    <label
                      class="chakra-form__label css-zecyy3"
                      for="field-:r6:"
                      id="field-:r6:-label"
                    >
                      End Date
                    </label>
                    <div
                      class="chakra-input__group css-bx0blc"
                      data-group="true"
                    >
                      <input
                        class="chakra-input css-1lx593h"
                        disabled=""
                        id="field-:r6:"
                        min=""
                        name="endDate"
                        type="date"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </section>
          </div>
        </div>
      </div>
      <button
        class="chakra-button css-4xx2wk"
        disabled=""
        type="submit"
      >
        Save
      </button>
    </div>
  </form>
</div>
`;
